File: golang/23_merge_k_lists/merge_k_lists.go
/*
 * @Date: 2023-08-12
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-08-12
 * @FilePath: /algorithm/golang/23_merge_k_lists/merge_k_lists.go
 */

// Package main ...
package main

import (
	"container/heap"
	"testing"

	"github.com/stretchr/testify/assert"
)

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeKLists(lists []*ListNode) *ListNode {
	pq := hp{}
	for _, head := range lists {
		if head != nil {
			pq = append(pq, head)
		}
	}
	heap.Init(&pq)
	dummy := &ListNode{}
	cur := dummy
	for len(pq) > 0 {
		cur.Next = heap.Pop(&pq).(*ListNode)
		cur = cur.Next
		if cur.Next != nil {
			heap.Push(&pq, cur.Next)
		}
	}
	return dummy.Next
}

type hp []*ListNode

func (h hp) Len() int           { return len(h) }
func (h hp) Less(i, j int) bool { return h[i].Val < h[j].Val }
func (h hp) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *hp) Push(v any)        { *h = append(*h, v.(*ListNode)) }
func (h *hp) Pop() any          { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v }

func main() {
	tests := []struct {
		lists []*ListNode
		ans   *ListNode
	}{
		{
			[]*ListNode{
				{1, &ListNode{4, &ListNode{5, nil}}},
				{1, &ListNode{3, &ListNode{4, nil}}},
				{2, &ListNode{6, nil}},
			},
			&ListNode{1, &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{4, &ListNode{5, &ListNode{6, nil}}}}}}}},
		},
		{
			[]*ListNode{},
			nil,
		},
		{
			[]*ListNode{nil},
			nil,
		},
	}
	t := &testing.T{}
	for _, item := range tests {
		assert.Equal(t, item.ans, item.lists)
	}
}

File: rust/23_merge_k_lists/merge_k_lists.rs
/*
 * @Date: 2023-08-12
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-08-12
 * @FilePath: /algorithm/rust/23_merge_k_lists/merge_k_lists.rs
 */

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

use std::{
    cmp::{Ord, PartialOrd, Reverse},
    collections::BinaryHeap,
};

impl Ord for ListNode {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.val.cmp(&other.val)
    }
}

impl PartialOrd for ListNode {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.val.cmp(&other.val))
    }
}

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut node = None;
        let mut cur = &mut node;
        let mut heap = BinaryHeap::new();
        for list in lists.into_iter() {
            if let Some(x) = list {
                heap.push(Reverse(x));
            }
        }
        while let Some(mut x) = heap.pop() {
            if let Some(y) = x.0.next.take() {
                heap.push(Reverse(y));
            }
            cur = &mut cur.insert(x.0).next;
        }
        node
    }
}

fn main() {
    let tests = vec![
        (
            vec![
                Some(Box::new(ListNode {
                    val: 1,
                    next: Some(Box::new(ListNode {
                        val: 4,
                        next: Some(Box::new(ListNode::new(5))),
                    })),
                })),
                Some(Box::new(ListNode {
                    val: 1,
                    next: Some(Box::new(ListNode {
                        val: 3,
                        next: Some(Box::new(ListNode::new(4))),
                    })),
                })),
                Some(Box::new(ListNode {
                    val: 2,
                    next: Some(Box::new(ListNode::new(6))),
                })),
            ],
            Some(Box::new(ListNode {
                val: 1,
                next: Some(Box::new(ListNode {
                    val: 1,
                    next: Some(Box::new(ListNode {
                        val: 2,
                        next: Some(Box::new(ListNode {
                            val: 3,
                            next: Some(Box::new(ListNode {
                                val: 4,
                                next: Some(Box::new(ListNode {
                                    val: 4,
                                    next: Some(Box::new(ListNode {
                                        val: 5,
                                        next: Some(Box::new(ListNode::new(6))),
                                    })),
                                })),
                            })),
                        })),
                    })),
                })),
            })),
        ),
        (vec![], None),
        (vec![None], None),
    ];

    for (lists, ans) in tests {
        assert_eq!(Solution::merge_k_lists(lists), ans);
    }
}
